### PR TITLE
bpo-40614: Respect feature version for f-string debug expressions

### PR DESCRIPTION
**Pull request by hauntsaninja**
https://bugs.python.org/issue40614

### CODE REVIEW

**Comment by lysnikolaou** [comment]
@pablogsal If this is okay, I guess we should also change the old parser to do this and backport that to 3.8 as well?

**Comment by pablogsal** [comment]
> I guess we should also change the old parser to do this and backport that to 3.8 as well?

Yeah, but the backport should be done manually to not include the parts for the new parser.

File: Lib/test/test_ast.py [hunk 1]
```diff
@@ -654,6 +654,11 @@ def test_ast_asdl_signature(self):
         expressions[0] = f"expr = {ast.expr.__subclasses__()[0].__doc__}"
         self.assertCountEqual(ast.expr.__doc__.split("\n"), expressions)
 
+    def test_issue40614_feature_version(self):
+        ast.parse('f"{x=}"', feature_version=(3, 8))
+        with self.assertRaises(SyntaxError) as e:
```
Comment (lysnikolaou) [inline Lib/test/test_ast.py]:
```suggestion
        with self.assertRaises(SyntaxError):
```

File: Misc/NEWS.d/next/Library/2020-05-18-22-41-02.bpo-40614.8j3kmq.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+:func:`ast.parse` will not parse self documenting expressions in f-strings when passed ``feature_version`` less than ``(3, 8)``.
```
Comment (lysnikolaou) [inline Misc/NEWS.d/next/Library/2020-05-18-22-41-02.bpo-40614.8j3kmq.rst]:
```suggestion
:func:`ast.parse` will not parse self documenting expressions in f-strings when passed ``feature_version`` is less than ``(3, 8)``.
```

File: Parser/pegen/parse_string.c [hunk 3]
```diff
@@ -931,6 +931,11 @@ fstring_find_expr(Parser *p, const char **str, const char *end, int raw, int rec
     /* Check for =, which puts the text value of the expression in
        expr_text. */
     if (**str == '=') {
+        if (p->feature_version < 8) {
+            RAISE_SYNTAX_ERROR("f-string: self documenting expressions are "
+                               "only supported in Python 3.8 and greater");
```
Comment (lysnikolaou) [inline Parser/pegen/parse_string.c:936]: Feel free to ignore this. (Note that if you don't the message should be changed in the old parser as well)
```suggestion
            RAISE_SYNTAX_ERROR("Self-documenting expressions in f-strings are "
                               "only supported in Python 3.8 and greater");
```

Comment (hauntsaninja) [reply Parser/pegen/parse_string.c:936]: Thanks! All the other f-string SyntaxErrors start with "f-string: ", so I think it's good to stay consistent.

**Comment by hauntsaninja** [comment]
Happy to open a PR with a manual backport using cherry_picker if this gets merged!

File: Parser/pegen/parse_string.c (hunk 3 above)
Comment (lysnikolaou) [reply Parser/pegen/parse_string.c:936]: Oh okay, didn't remember correctly then.

**Review by lysnikolaou — APPROVED** [review]
Thanks!

**Comment by hauntsaninja** [comment]
Should that label be "needs backport to 3.8"?

**Comment by lysnikolaou** [comment]
Nope, `master` is 3.10 as of about 3 hours ago. This needs to be backported to 3.9 as-is and then manually also to 3.8.

**Review by pablogsal — APPROVED** [review]

**Comment by hauntsaninja** [comment]
Opened a backport to 3.8 in #20466

**Pull request merged**